RFC.md

```diff
@@ -0,0 +1,413 @@
+        pos_embeddings: Optional[nn.Module] = None,
+        # support qk normalization
+        # https://arxiv.org/abs/2010.04245
+        q_norm: Optional[nn.Module] = None,
+		k_norm: Optional[nn.Module] = None,
+		# sets the default mask wehn none is provided
+		# to be causal or None
+        default_causal_mask: bool = True,
+    ) -> None:
+        super().__init__()
+        ...
+        if bool(q_norm) ^ bool(k_norm):
+            raise ValueError("q and k norm must be set together")
+		...
+
+    def forward(
+        self,
+        x: Tensor,
+        # Optinal for cross attention but we could
+        # make it required instead and self attention
+        # would be GQA(x, x)
+        y: Optional[Tensor] = None,
+        *,
+        mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+        ...
+        bsz, seq_len_x, _ = x.shape
+        y = y if y is not None else x
+        _, seq_len_y, _ = y.shape
+
+        # q has shape [b, s, num_heads * head_dim]
+        # k has shape [b, s, num_kv_heads * head_dim]
+        # v has shape [b, s, num_kv_heads * head_dim]
+        q = self.q_proj(x)
+        k = self.k_proj(y)
+        v = self.v_proj(y)
+
+        ...
+
+        # Apply positional embeddings
+        if self.pos_embeddings is not None:
+            q = self.pos_embeddings(q, input_pos=input_pos)
+            k = self.pos_embeddings(k, input_pos=input_pos)
+
+        ...
+
+        # Normalize k and q
+        if self.q_norm is not None:
+            q = self.q_norm(q)
+            k = self.k_norm(k)
+
+        ...
+
+        output = nn.functional.scaled_dot_product_attention(
+            q,
+            k,
+            v,
+            attn_mask=mask,
+            dropout_p=self.attn_dropout,
+            # if you're using a kv_cache or a mask is provided
+            # then don't apply a default causal mask
+            # if there is no mask,
+            # then only apply a causal mask if GQA has default_causal_mask
+            is_causal=self.kv_cache is None and mask is None and self.is_causal,
+        )
+
+        ...
+
+        return self.output_proj(output)
+```
+
+### Legacy Checkpointing
+
+Since GQA changes sa_norm to attn_norm, this will break existing checkpoints that are in the tune format. It could be argued that since we don't really let users save checkpoints in the tune format by default, that this would not affect anyone. But to ensure that no checkpoint are broken, we can add conversion mappings to [convert_weights.py](https://github.com/pytorch/torchtune/blob/main/torchtune/models/convert_weights.py). These would be exactly the same as `_FROM_HF` or `_FROM_META` but instead map from older version to new.
+
+```python
+_FROM_TUNE_0_2_1 = {
+    "layers.{}.sa_norm.scale": "layers.{}.attn_norm.scale",
+}
+```
+The above code block is assuming that the mapping function is updated to treat missing keys as an identity
```

Comment:
This is a secondary point to some of the other things in this RFC but I might reconsider this. It is probably best to be explicit about our conversions, especially with respect to external checkpoint formats (e.g. if some new key is added and we assume the identity mapping it can get tricky to debug). We can also consider our internal torchtune checkpoint versioning as a separate API, in which case we can assume the identity mapping for our own internal versioning but not for any other checkpoint remappings.

Reply:
Should we save a torchtune version or hash in each model config?